### PR TITLE
feat(form): add transformErrors prop

### PR DESCRIPTION
**Pull request by johannesss**
Make it possible to provide the [transformErrors](https://react-jsonschema-form.readthedocs.io/en/latest/api-reference/form-props/#transformerrors) function to customize the error messages.

## Review:
- [x] Commits are [atomic](https://seesparkbox.com/foundry/atomic_commits_with_git)
- [x] Commits have the correct *type* for the changes made
- [x] Commits with *breaking changes* are marked as such

### Browsers tested:
(Check any that applies, it's ok to leave boxes unchecked if testing something didn't seem relevant.)

Windows:
- [ ] Chrome
- [ ] Edge
- [ ] Firefox

Linux:
- [ ] Chrome
- [ ] Firefox

macOS:
- [x] Chrome
- [ ] Firefox
- [ ] Safari

Mobile:
- [ ] Chrome on Android
- [ ] iOS

### CODE REVIEW

File: src/components/form/form.tsx
Comment (adrianschmidt) [inline src/components/form/form.tsx:81]: We don't want to expose types from `rjsf` to our consumers, as that locks us into `rjsf`, it also means that updating `rjsf` could become more difficult if there are breaking changes made to the `AjvError` interface. Can we create a proxy interface, that exposes reasonably generic properties, likely to still be available (or made available through an adapter) if the underlying implementation changes or is swapped out?

Comment (johannesss) [reply src/components/form/form.tsx]: I've made some changes based on your comment

Comment (adrianschmidt) [reply src/components/form/form.tsx]: Nice! I'll have a look 👍

File: src/components/form/form.tsx
Comment (adrianschmidt) [inline src/components/form/form.tsx:248]: Oooh, so nice! Both with the hiding of AjvError and the example you added 😍 👏👏👏

File: src/components/form/examples/custom-error-message-form.tsx
Comment (adrianschmidt) [inline src/components/form/examples/custom-error-message-form.tsx:1]: Small feedback on the commit messages:

### First commit:

> feat: add transformErrors prop

- add a "scope" when applicable: `feat(form): …`
This can sometimes be a little arbitrary, but here in Lime Elements it's almost always super easy, because we use the component's name, minus the `limel-` prefix. In other repos it's often a bit more complicated, since we tend do have deeper structures, and you might want to keep the "scope" pretty short.
- write the commit message subject so it makes sense for someone reading the changelog
The first line of the commit message (`type(scope): subject`) is used to create the changelog entry, so use a scope and a subject that makes sense in that context. Try to summarise what has been done from a user's perspective when possible. The first line can be up to 100 characters, including the type and scope, so if it fits you can also include something about what problem it fixes. In this case, you already have a solid summary in the commit message body, although I'd actually skip the "provided by JSON schema" part: `feat(form): make it possible to customize the default error messages` 👌
- include `fix: #123` at the bottom of the commit message (if the commit fixes an issue)
That way, the changelog item for this commit will include a link directly to the issue it fixes, like for the second item here: image

---

### The last three commits

The last three commits should just be fixup commits, since they fix things in a commit that hasn't been merged to `next. Fixup commits indicate that they should be squashed into an existing commit. We also have a PR check that fails if there are any fixup commits. That lets us make any changes based on review feedback with fixup commits, that remain un-squashed while the review is ingoing (so we can review only the changes each time something new is pushed), while the PR check prevents us from accidentally merging the PR before squashing the fixups. [Here's a good post](https://thoughtbot.com/blog/autosquashing-git-commits) explaining fixup commits better than I can.

---

I can help you out with editing the existing commits, if you want me to 👍

File: src/components/form/form.tsx
Comment (johannesss) [reply src/components/form/form.tsx:248]: Thanks a lot @adrianschmidt!

File: src/components/form/examples/custom-error-message-form.tsx
Comment (johannesss) [reply src/components/form/examples/custom-error-message-form.tsx:1]: Thank you for the feedback @adrianschmidt, it all makes sense! I will review the guide you linked and conform to these rules in future PRs!

Comment (johannesss) [reply src/components/form/examples/custom-error-message-form.tsx:1]: Yes, some help with editing the existing commits would be great!

Comment (adrianschmidt) [reply src/components/form/examples/custom-error-message-form.tsx:1]: Fixed! Since the commit history was the last thing to fix before I'd approve and merge this, I realised it was easiest to just squash the commits while I was at it.

(Just to clarify: we don't just blindly squash each PR to _one_ commit. We try to make [atomic commits](https://github.com/Lundalogik/lime-elements/blob/next/CONTRIBUTING.md#atomic-commits), and consider tests, docs, and examples for a new feature as the same "atom" as the feature itself. So in this case, a single commit 🙂)

**Review by adrianschmidt — APPROVED** [review]

**Review by adrianschmidt — CHANGES_REQUESTED** [review]
Hang on 😄

File: src/components/form/form.tsx
Comment (adrianschmidt) [inline src/components/form/form.tsx:107]: I don't think we should bind this function to `this`. `this.transformErrors` is an external function, and binding it to `this` will give it access to the internals of the form component. Although, I guess, since it's called as a method of `this` it might already have that access… 🤔

Regardless, I don't think we need this bind, and if we don't we should remove it.

Comment (johannesss) [reply src/components/form/form.tsx:107]: Hmm, I'll have a look at this!

File: src/components/form/form.tsx
Comment (adrianschmidt) [inline src/components/form/form.tsx:106]: We can also get rid of this bind, by using an arrow function instead. We try to use arrow functions instead of binds when possible nowadays, but there's a bunch of old code still using bind, because we used that a lot a few years ago.

Comment (adrianschmidt) [reply src/components/form/form.tsx:106]: Although, you don't really have to change that now, because once we've merged this, I'll change the already existing functions in this file anyway, in a separate PR, so I can do this one then too.

Comment (johannesss) [reply src/components/form/form.tsx:106]: Ahh yes, you're right! I will commit some changes shortly @adrianschmidt

File: src/components/form/form.tsx
Comment (adrianschmidt) [reply src/components/form/form.tsx:107]: I had to look into whether internals are actually exposed, and indeed they are. I have a fix, that I can push if ok with you?

File: src/components/form/form.tsx
Comment (johannesss) [reply src/components/form/form.tsx:106]: Sounds good! @adrianschmidt

File: src/components/form/form.tsx
Comment (johannesss) [reply src/components/form/form.tsx:107]: Yes of course, please do!

File: src/components/form/form.tsx
Comment (adrianschmidt) [inline src/components/form/form.tsx:268]: @jgroth I think we probably have more places where we call a consumer supplied callback as a method on `this`, and we need to use this pattern on every single one of them, or we have a security problem. I don't think we need to go find every instance right now, but we should at least keep this in the back of our heads and make sure to fix existing instances when we come across them, and also not let new instances in.

I've tested what happens if the callback is created using an arrow function, or if the consumer uses `bind(this)` before sending in the callback to limel-form, and that still works as expected—the `.call({}, …)` doesn't destroy the existing binding.

cc @BregenzerK FYI

File: src/components/form/form.tsx
Comment (adrianschmidt) [reply src/components/form/form.tsx]: ⚡

Comment (adrianschmidt) [reply src/components/form/form.tsx]: Fixed. I'll squash and approve, if you're happy?

Comment (johannesss) [reply src/components/form/form.tsx]: Great, looks good to me!

Comment (johannesss) [reply src/components/form/form.tsx]: I tried to come up with a different prop name for `limel-form` than `transformErrors` to differentiate from rjsf, but failed. It feels clear concise to me, but feel free to change it.

Comment (adrianschmidt) [reply src/components/form/form.tsx]: Nah, I think `transformErrors` is great. No need to _differentiate_ from rjsf, just to _decouple_ 👍

**Review by adrianschmidt — APPROVED** [review]

File: src/components/form/form.tsx
Comment (jgroth) [reply src/components/form/form.tsx:268]: I don't think there are that many places actually :slightly_smiling_face: There's the `propFactory` that's also part of the form component and the `searcher` that's part of the picker. Besides those I don't think we have props that are functions. We might have some that are part of an inner structure of some prop, but I doubt we would bind it to the component in those cases :sweat_smile: 

Another way to get rid of the `this` binding is to simply assign the function to a constant/variable before invoking it, e.g.
```ts
const transformErrors = this.transformErrors;
transformErrors();
```

Comment (johannesss) [reply src/components/form/form.tsx:268]: Oh, that's a nifty solution. Nice!

Comment (BregenzerK) [reply src/components/form/form.tsx:268]: Sorry, I need some more context here. Is there anything we need to check and refactor in our add-ons?

Comment (adrianschmidt) [reply src/components/form/form.tsx:268]: @BregenzerK Sorry, I mentioned you since you're a Lime Elements maintainer. I should have mentioned the group instead, to make the intention clear. I don't think you need to worry about this in add-ons, no.

Comment (adrianschmidt) [reply src/components/form/form.tsx:268]: > I don't think there are that many places actually 🙂 There's the `propFactory` that's also part of the form component and the `searcher` that's part of the picker. Besides those I don't think we have props that are functions. We might have some that are part of an inner structure of some prop, but I doubt we would bind it to the component in those cases 😅

I agree we probably don't have many. We might have none actually, when I think about it. But we should at least keep in mind that this pattern could be dangerous, and that doing a `bind(this)` isn't necessary to open the vulnerability; it's enough to have a callback sent in as a property, and then calling it directly from `this` (like `this.consumerProvidedCallback()` for example).

> Another way to get rid of the `this` binding is to simply assign the function to a constant/variable before invoking it, e.g.
> 
> ```ts
> const transformErrors = this.transformErrors;
> transformErrors();
> ```

Nice 👍

Comment (adrianschmidt) [reply src/components/form/form.tsx:268]: > > Another way to get rid of the `this` binding is to simply assign the function to a constant/variable before invoking it, e.g.
> > ```ts
> > const transformErrors = this.transformErrors;
> > transformErrors();
> > ```
> 
> Nice 👍

This one might run a greater risk of being "refactored" into the vulnerable version though… 🤔

It's pretty easy to write a test to check that the provided callback doesn't have access to the inner scope though, and as long as we do that, I like your suggestion 👍

**Pull request merged**
